pymc_marketing/mmm/models/components/lagging.py

```diff
@@ -0,0 +1,394 @@
+    A class representing a Weibull CDF adstock component in a marketing mix model.
+
+    Parameters:
+    -----------
+    max_lagging : int
+        The maximum lagging value for the adstock component.
+    model : pymc.Model
+        The PyMC model object.
+    model_config : dict
+        The configuration dictionary for the model.
+
+    Attributes:
+    -----------
+    REQUIRED_KEYS : list
+        The list of required keys in the model configuration.
+
+    Methods:
+    --------
+    apply(data: Union[pm.Data, pm.MutableData]) -> pm.Deterministic:
+        Apply the Weibull CDF adstock component to the given data.
+
+    """
+
+    REQUIRED_KEYS = ["adstock_lambda", "adstock_shape"]
+
+    def __init__(
+        self,
+        max_lagging: int = 4,
+        model: Optional[pm.Model] = None,
+        model_config: Optional[Dict] = None,
+    ):
+        """
+        Initialize the WeibullCDFAdstockComponent.
+
+        Parameters:
+        -----------
+        max_lagging : int
+            The maximum lagging value for the adstock component.
+        model : pymc.Model
+            The PyMC model object.
+        model_config : dict
+            The configuration dictionary for the model.
+
+        """
+        self.max_lagging = max_lagging
+        self.model = model
+        if model_config is not None and not all(
+            key in model_config for key in self.REQUIRED_KEYS
+        ):
+            self.model_config = {**model_config, **self._default_lagging_config}
+        elif model_config is None:
+            self.model_config = self._default_lagging_config
+        else:
+            self.model_config = model_config
+
+    @property
+    def _default_lagging_config(self) -> Dict[str, Dict]:
+        return {
+            "adstock_lambda": {
+                "dist": "Beta",
+                "kwargs": {"alpha": 1, "beta": 3},
+            },
+            "adstock_shape": {
+                "dist": "Beta",
+                "kwargs": {"alpha": 1, "beta": 3},
+            },
+        }
+
+    @property
+    def variable_mapping(self) -> Dict[str, str]:
+        """The mapping from saturation function parameters to variables in a model."""
+        return {
+            "lam": "adstock_lambda",
+            "k": "adstock_shape",
+        }
+
+    def apply(self, data: Union[pm.Data, pm.MutableData]) -> pm.Deterministic:
+        """
+        Apply the Weibull CDF adstock component to the given data.
+
+        Parameters:
+        -----------
+        data : Union[pm.Data, pm.MutableData]
+            The data to apply the adstock component to.
+
+        Returns:
+        --------
+        pm.Deterministic
+            The adstocked data as a PyMC Deterministic variable.
+
+        """
+        self.adstock_lambda_dist = _get_distribution(
+            dist=self.model_config["adstock_lambda"]
+        )
+
+        self.adstock_shape_dist = _get_distribution(
+            dist=self.model_config["adstock_shape"]
+        )
```

Comment:
Would these be used anywhere else that is not this method? If so, just using variables that are local to this method makes the most sense to me